File: Stack/InfixEvaluation.py
from collections import deque
def precedence(character):
    if character=='-' or character== '+':
        return 1
    elif character== '*' or character=='/':
       return 2


def operation(val1,val2,operator):
    if(operator=='+'):
        return val1+val2

    elif(operator=='-'):
        return val1-val2

    elif(operator=='*'):
        return val1*val2

    elif(operator=='/'):
        return val1/val2



def infix(exp):
    operators=deque()
    operands=deque()
    for char in exp:
        if char=='(':
            operators.append(char)
        elif char.isdigit():
            operands.append(int(char))
        elif char=='+' or char=='-' or char=='*' or char =='/':
            while len(operators)>0 and operators[-1]!='(' and precedence(char)<=precedence(operators[-1]):
                val2 = operands.pop()
                val1 = operands.pop()
                op =  operators.pop()
                opval = operation(val1, val2, op)
                operands.append(opval)
            operators.append(char);
        elif char==')':
            while len(operators)>0 and operators[-1]!='(':
                val2 = operands.pop()
                val1 = operands.pop()
                op =  operators.pop()
                opval = operation(val1, val2, op)
                operands.append(opval)
            if len(operators)>0:
                operators.pop()
        


    while len(operators)>0:
        val2 = operands.pop()
        val1 = operands.pop()
        op =  operators.pop()
        opval = operation(val1, val2, op)
        operands.append(opval)
    
    val=operands.pop()
    print(round(val))
    



    
exp = input()
exp = exp.replace(" ", "")


infix(exp)

# i/p test case
# 2 + 6 * 4 / 8 - 3

File: Stack/Stack-Java/InfixEvaluation.java
import java.io.*;
import java.util.*;

public class Main{

    public static int precedance(char op){
    if(op == '+'){
      return 1;
    } else if(op == '-'){
      return 1;
    } else if(op == '*'){
      return 2;
    } else {
      return 2;
    }
  }

  public static int operation(int val1, int val2, char op){
    if(op == '+'){
      return val1 + val2;
    } else if(op == '-'){
      return val1 - val2;
    } else if(op == '*'){
      return val1 * val2;
    } else {
      return val1 / val2;
    }
  }
  

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String exp = br.readLine();
    
        Stack<Integer> operands=new Stack<Integer>();
        Stack<Character> operators=new Stack<Character>();
        for(int i=0;i<exp.length();i++){
            char ch=exp.charAt(i);
            if(Character.isDigit(ch)){
                operands.push(ch-'0');
            }
            else if(ch=='('){
                operators.push(ch);
    
            }
    
            else if(ch==')'){
                while(operators.size() > 0 && operators.peek()!='('){
                    char op=operators.pop();
                    int b =operands.pop();
                    int a =operands.pop();
                    int opv=operation(a,b,op);
                    operands.push(opv);
                }
    
                if (operators.size() > 0) {
                      operators.pop();
                }
                
            }
    
            else if(ch=='+' || ch=='-' || ch=='/' || ch=='*'){
                while(operators.size()>0 && operators.peek()!='(' 
                      && precedance(ch)<=precedance(operators.peek())){
                    char op=operators.pop();
                    int b =operands.pop();
                    int a =operands.pop();
                    int opv=operation(a,b,op);
                    operands.push(opv);
    
                }
                operators.push(ch);
                
            }
        }
    
        System.out.println(operands.peek());
       
     }
}
